docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+
+Language support for [`AsyncSequence`][SE-0298] also allows for writing functions that return values over time.
+
+We would like to explore how we could offer APIs that make use of these new language features to allow users to run HTTPRequest using these new idioms.
+
+This proposal describes what these APIs could look like and explores some of the potential usability concerns.
+
+## Proposed API additions
+
+### New `AsyncRequest` type
+
+The proposed new `AsyncRequest` shall be a simple swift structure.
+
+```swift
+struct AsyncRequest {
+  /// The requests url.
+  var url: String
+  
+  /// The request's HTTPMethod
+  var method: HTTPMethod
+  
+  /// The request's headers
+  var headers: HTTPHeaders
+
+  /// The request's body
+  var body: Body?
+    
+  init(url: String) {
+    self.url = url
+    self.method = .GET
+    self.headers = .init()
+    self.body = .none
+  }
+}
+```
+
+A notable change from the current [`HTTPRequest`][HTTPRequest] is that the url is not of type `URL`. This makes the creation of a request non throwing. Existing issues regarding current API: 
+
+- [HTTPClient.Request.url is a let constant][issue-395]
+- [refactor to make request non-throwing](https://github.com/swift-server/async-http-client/pull/56)
+- [improve request validation](https://github.com/swift-server/async-http-client/pull/67)
+
+The url validation will become part of the normal request validation that occurs when the request is scheduled on the `HTTPClient`. If the user supplies a request with an invalid url, the http client, will reject the request.
+
+In normal try/catch flows this should not change the control flow:
+
+```swift
+do {
+  var request = AsyncRequest(url: "invalidurl")
+  try await httpClient.execute(request, deadline: .now() + .seconds(3))
+} catch {
+  print(error)
+}
+```
+
+If the library code throws from the AsyncRequest creation or the request invocation the user will, in normal use cases, handle the error in the same catch block. 
+
+#### Body streaming
+
+The new AsyncRequest has a new body type, that is wrapper around an internal enum. This allows us to evolve this type for use-cases that we are not aware of today. 
+
+```swift
+public struct Body {
+  static func bytes<S: Sequence>(_ sequence: S) -> Body where S.Element == UInt8
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == ByteBuffer
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == UInt8
+}
+```
+
+The main difference to today's `Request.Body` type is the lack of a `StreamWriter` for streaming scenarios. The existing StreamWriter offered the user an API to write into (thus the user was in control of when writing happened). The new `AsyncRequest.Body` uses `AsyncSequence`s to stream requests. By iterating over the provided AsyncSequence, the HTTPClient is  in control when writes happen, and can ask for more data efficiently. 
```

Comment:
We don't lose this capability actually, we just express it the right way (well, "right" in my brain at least) now :-)

One should always accept "some" async sequence. The way this sequence is produced matters not to the consumer of it (this API).

Now, the way a "writer" is expressed, is by the end-user creating an `Swift.AsyncStream` (which actually is more like "async sequence factory" though the name stuck). That stream then can be yielded into: `stream.yield(some bytes) -> YieldResult` and this way the end user may construct their own "writer style" API, and we don't have to prepare anything for it, since we only deal with it in terms of `AsyncSequence` 👍

Reply:
We do lose it: `AsyncStream.yield` is not isomorphic to `StreamWriter`. Importantly, `StreamWriter.write` _returns an `EventLoopFuture`_, which allows it to exert backpressure on the producer. `AsyncStream.Continuation.yield` is not an async function, and so it cannot exert backpressure in the same way. Users could retrofit it in a ghetto kinda way by using a buffered stream and observing the return values from `yield`. They can also retrofit it by using `AsyncStream(unfolding:onCancel:)` but that forces them to construct a closure and essentially "invert" their backpressure (from push-based to pull-based) which is fairly non-obvious.

gRPC is proposing an alternative API that keeps the `StreamWriter`-like design, which tbh I prefer.

Reply:
That's the entire reason behind the yielded values which we made sure to include for exactly this reason.

You're saying that it's dependent on the intermediate buffer being full or not, and not on the actual "end that does the write being full or not" but that's just the reality of dealing with buffered streams -- and _any_ stream may be buffered so that's it, you won't get the exact same semantics but these are the semantics you work with in stream APIs.

Reply:
There is a minor issue implementing `StreamWriter` using `AsyncStream`. If I am looking for the following api
```swift
let stream = StreamWriter()
stream.write(buffer1)
stream.write(buffer2)
stream.write(nil)
```
I need the `StreamWriter` to be ready to accept ByteBuffers immediately. The construction of `AsyncStream` takes a `build` closure where you setup your stream. The api does not make it obvious that this closure will have been called by the time `AsyncStream.init` has finished. Having looked at the code I know this is the case but can I assume that will always be the case. Here is the best I could do to implement above
```swift
class StreamWriter {
    var write: (ByteBuffer?) -> ()
    let stream: AsyncStream<ByteBuffer>
    
    init() {
        var write: ((ByteBuffer?) -> ())?
        self.stream = .init { cont in
            write = { buffer in
                guard let buffer = buffer else { cont.finish(); return }
                cont.yield(buffer)
            }
        }
        self.write = write!
    }
}
```
I had to force unwrap the `write` closure which makes me unhappy.

Reply:
I assume we'd fix the `StreamWriter` API, right? There's already  #291

Reply:
problem statements in #194 and #264

Reply:
Aren't we discussing removing it and replacing with `AsyncSequence` which in the end I'm all for. I do think supplying a stream implementation with back pressure would be helpful as `AsyncStream` isn't up to the job yet.

Reply:
We know about that limitation and now ticketified the specific case for  it: https://bugs.swift.org/browse/SR-15076